### DISCUSSION TITLE
OpenID Connect - OIDC with Azure AD issues

### DISCUSSION THREAD

**Discussion by vb-03** [pretix/pretix#3130] [Q&A]
Hi,
I'm experimenting with customer accounts using OpenID. I'm using a Azure AD tenant, and set-up the SSO provider.
However, when logging in, everything goes smoothly in Azure side but then I get an Internal Server Error, and there's this in the logs:
![image](https://user-images.githubusercontent.com/47537857/219803068-c57f7295-4203-4c58-a07a-0298bb38fe92.png)
It could totally be an error in how I set things up in the Azure side, but this error isn't giving any clarity...
Any ideas?

**Comment by raphaelm** [comment]
It looks like the OpenID `state` parameter that we pass to the provider on the first redirect doesn't get passed back to pretix exactly on the last redirect – can you check that theory through looking at the URLs in your browser dev tools?

**Reply by vb-03** [reply]
First redirect (the green underline is the organizer Short form)
![image](https://user-images.githubusercontent.com/47537857/219815060-8d4c36d9-dd3b-429f-afee-c70ba375cc02.png)

The last redirect from MS side, it seems you're right
![image](https://user-images.githubusercontent.com/47537857/219815153-6593e909-eb5b-44f3-bf4b-052d1de35b64.png)

EDIT: Fixed 2nd picture

**Reply by raphaelm** [reply]
sorry, by last redirect I meant the one from Microsoft back to pretix

**Reply by vb-03** [reply]
Oh, my bad, misunderstood

![image](https://user-images.githubusercontent.com/47537857/219815930-bb444c84-fa80-4399-8bc9-6766ea24f2bf.png)

It is there...

**Reply by vb-03** [reply]
There is a mismatch between the sent state (1st) and the received one (2nd) tho
![image](https://user-images.githubusercontent.com/47537857/219816253-b88ac3a7-ac47-40c9-9b6f-39e9a7b95849.png)

**Reply by vb-03** [reply]
I've been debugging this in my Pretix Dev, it seems to be an issue getting either the nonce or the redirect url, still unsure what as everything is new to me..
It only gets the "state" part of the URL, which I'm assuming is taking the place of the nonce?

**Reply by raphaelm** [reply]
We're misusing the "state" a little to be both a csrf protection token *and* carrier of the redirect URL, separated by a separation character. This is already the second character we try for this, we might need to find a different approach in the future

**Reply by raphaelm** [reply]
Can you test if https://github.com/pretix/pretix/pull/3222 fixes this for you?

**Reply by vb-03** [reply]
It works now!
Tested with regular Azure.
Testing next with Azure B2C which is more costumer-focused, however I'll need a bit more time due to the set-up of the UserInfo endpoint there (in B2C isn't supported by default - https://help.salesforce.com/s/articleView?id=000393769&type=1)
